lib/cocoapods/downloader/cache.rb

```diff
@@ -69,6 +69,93 @@ def cache_descriptors_per_pod
         end
       end
 
+      # Convenience method for acquiring a shared lock to safely read from the
+      # cache. See `Cache.lock` for more details.
+      #
+      # @param  [Pathname] location
+      #         the path to require a lock for.
+      #
+      # @param  [Proc] block
+      #         the block to execute inside the lock.
+      #
+      # @return [void]
+      #
+      def self.read_lock(location, block = Proc.new)
+        Cache.lock(location, File::LOCK_SH, block)
+      end
+
+      # Convenience method for acquiring an exclusive lock to safely write to
+      # the cache. See `Cache.lock` for more details.
+      #
+      # @param  [Pathname] location
+      #         the path to require a lock for.
+      #
+      # @param  [Proc] block
+      #         the block to execute inside the lock.
+      #
+      # @return [void]
+      #
+      def self.write_lock(location, block = Proc.new)
+        Cache.lock(location, File::LOCK_EX, block)
+      end
+
+      # Creates a .lock file at `location`, aquires a lock of type
+      # `lock_type`, checks that it is valid, and executes `block` while holding
+      # on to that lock. Afterwards, the .lock file is deleted, which is why
+      # validation of the lock is necessary, as you might have a lock on a file
+      # that doesn't exist on the filesystem anymore.
+      #
+      # @param  [Pathname] location
+      #         the path to require a lock for.
+      #
+      # @param  [locking_constant] lock_type
+      #         the type of lock, either exclusive (File::LOCK_EX) or shared
+      #         (File::LOCK_SH).
+      #
+      # @param  [Proc] block
+      #         the block to execute inside the lock.
+      #
+      # @return [void]
+      #
+      def self.lock(destination, lock_type, block = Proc.new)
+        lockfile = "#{destination}.lock"
+        f = nil
+        loop do
+          f.close if f
+          f = File.open(lockfile, File::CREAT, 0o644)
+          f.flock(lock_type)
+          break if Cache.validate_lock(f, lockfile)
+        end
+        begin
+          block.call destination
+        ensure
+          if lock_type == File::LOCK_SH
+            f.flock(File::LOCK_EX)
+            File.delete(lockfile) if Cache.validate_lock(f, lockfile)
+          else
+            File.delete(lockfile)
+          end
+          f.close
+        end
+      end
+
+      # Checks that the lock is on a file that still exists on the filesystem.
+      #
+      # @param  [File] file
+      #         the actual file that we have a lock for.
+      #
+      # @param  [String] filename
+      #         the filename of the file that we have a lock for.
+      #
+      # @return [Boolean]
+      #         true if `filename` still exists and is the same file as `f`
+      #
+      def self.validate_lock(file, filename)
```

Comment:
Perhaps slightly more Rubyesque; nothing important, though.

```suggestion
      def self.has_lock?(file, filename)
```

Reply:
Well, it has a lock, the question is if the lock is on a file that exists on the filesystem. I suppose `valid_lock?` might be a better name, though.